lib/view_component/compiler.rb

```diff
@@ -163,9 +163,8 @@ def matching_views_in_source_location
       # end
       #
       # Without this, `MyOtherComponent` will not look for `my_component/my_other_component.html.erb`
-      nested_component_files = if component_class.name.include?("::")
-        nested_component_path = component_class.name.deconstantize.underscore
-        Dir["#{directory}/#{nested_component_path}/#{component_name}.*{#{extensions}}"]
+      nested_component_files = if component_class.name.include?("::") && component_name != filename
+        Dir["#{directory}/#{filename}/#{component_name}.*{#{extensions}}"]
```

Comment:
I'm definitely looking for another set of 👀 here, as this is not the simplest piece of code ever written.

Reply:
This will work for components that are nested one level deep, but I *think* it might break if they're nested any deeper than that since we're relying on `filename`.

e.g.

```ruby
# app/components/foo.rb
class Foo < ViewComponent::Base
  class Bar < ViewComponent::Base
    class Baz < ViewComponent::Base
    end
  end
end
```

I think that would look for `Baz`s template in `app/components/foo/baz.html.erb`, when I think we'd want `app/components/foo/bar/baz.html.erb`.

Can we add a test for this case?

Reply:
Ahh yeah you're totally right @BlakeWilliams!

Part of me actually is inclined to have this behave exactly as it does with this change, essentially flattening the component hierarchy, but I can see how that could be surprising.

I'm curious if you have an idea for how we might cleanly handle the nested scenario, b/c I'm struggling to think of a good solution.

Reply:
@BlakeWilliams the more I think about this, the more I'm inclined to have this be the "correct" behavior, even when a component is deeply nested. That keeps the file system layout identical for component files and template files, and happens to keep the template lookup relatively simple. I've updated the test case to at least cover the scenario.

If we *did* want to somehow look for deeply nested component template files differently, I think we'd have to somehow walk [`module_parents`](https://api.rubyonrails.org/v6.0.3.4/classes/Module.html#method-i-module_parents), determine if each parent is a `ViewComponent::Base`, compare each one's `source_location`, and then build up the path that way.

If others disagree, I can definitely try and tackle the deeply nested scenario differently.

Reply:
> That keeps the file system layout identical for component files and template files, and happens to keep the template lookup relatively simple. I've updated the test case to at least cover the scenario.

That makes sense to me. I think a good thing to follow here might be how controllers lookup their template paths. If we had a controller that was nested inside of another controller, would this be consistent with how its lookup paths behave?

Reply:
Meaning something like this?

```ruby
class FooController < ApplicationController
  def index
  end

  class BarController < ApplicationController
    def index
    end
  end
end
```

When giving this a shot in an app locally, Rails isn't able to look up the controller successfully, let alone look up the template, so it might not be something we can mimic.

Reply:
Yeah, I was curious if Rails handled that situation. Thanks for trying that out though!

I've thought about this a bit and I think you're right that this is the way to go. 👍